### PR TITLE
Introduces member bouncing test support

### PR DESCRIPTION
**Pull request by vbekiaris**
* Introduces `BounceMemberRule` JUnit rule to facilitate testing behavior while members in the cluster are being stopped & started.
* Member and client-side querying during migrations tests are migrated to use `BounceMemberRule`

### CODE REVIEW

**Comment by vbekiaris** [comment]
Failed test: `com.hazelcast.internal.nearcache.impl.invalidation.MetaDataGeneratorTest.destroying_map_removes_related_metadata`

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by tombujok** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/test/bounce/BounceMemberRule.java [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+    }
+
+    // wait until test tasks complete or throw exception
+    private void waitForFutures(Future[] futures) {
+        boolean hasTestFailed = testFailed.get();
+        for (Future future : futures) {
+            try {
+                // TODO check logic here
+                // if the test failed, try to locate immediately the future that is done and will throw an exception
+                if ((hasTestFailed && future.isDone()) || !hasTestFailed) {
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/test/bounce/BounceMemberRule.java]: this indeed looks suspicious

Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/test/bounce/BounceMemberRule.java]: I updated the future-waiting code to avoid the edge case where `testFailed.get()` might return `true` while its `Future` might not be done yet.
